### ISSUE TITLE
--ignorePattern option is confusing

### ISSUE THREAD

**Issue by bodote** [stryker-mutator/stryker-js#3688]
**Summary**

- the `--file` AND `--ignorePatterns` option fails , the due to "Unable to find karma config at 'karma.conf.js'"  in the sandbox ERROR   and 
- although there is **at least one** "Files to mutate" is found , "Instrumenter Instrumented 1 source file(s) with 28 mutant(s)" but then
- "Chrome Headless 104.0.5112.101 (Mac OS 10.15.7): **Executed 0 of 0 SUCCESS** (0 secs / 0 secs)"
- however: stryker runs without problems without the `--file `or the `--ignorePatters` option.
- however 2 : the `--file` option was working yesterday, before  I upgraded Java from 17 to 18 and rebooted of my Macbook, but that should be totally unrelated, shouldn't it ? 

if I start stryker on ALL  files of the project, the karma config at 'karma.conf.js'" is found in the sandbox, just as expected , according to the stryker logfile.

see project : 
https://github.com/bodote/book-monkey/commit/c8e53ce1be19377f9409515923b2406fd233004b

**Stryker config**

```json
{
  "$schema": "./node_modules/@stryker-mutator/core/schema/stryker-schema.json",
  "_comment": "This config was generated using 'stryker init'. Please see the guide for more information: https://stryker-mutator.io/docs/stryker-js/guides/angular",
  "mutate": [
    "src/**/*.ts",
    "!src/**/*.spec.ts",
    "!src/test.ts",
    "!src/environments/*.ts"
  ],
  "testRunner": "karma",
  "karma": {
    "configFile": "karma.conf.js",
    "projectType": "angular-cli",
    "config": {
      "browsers": [
        "ChromeHeadless"
      ]
    }
  },

  "reporters": [
    "progress",
    "clear-text",
    "html"
  ],
  "concurrency": 4,
  "concurrency_comment": "Recommended to use about half of your available cores when running stryker with angular",
  "coverageAnalysis": "perTest",
  "timeoutFactor": 40,
  "checkers": ["typescript"],
  "tsconfigFile": "tsconfig.json",
  "ignorePatterns": ["**","!src/**/book-details*.ts"],
  "logLevel": "trace"
}
```

**Test runner config**

**Stryker environment**
```javascript
// Karma configuration file, see link for more information
// https://karma-runner.github.io/1.0/config/configuration-file.html

module.exports = function (config) {
  config.set({
    basePath: '',
    frameworks: ['jasmine', '@angular-devkit/build-angular'],
    plugins: [
      require('karma-jasmine'),
      require('karma-chrome-launcher'),
      require('karma-jasmine-html-reporter'),
      require('karma-coverage'),
      require('@angular-devkit/build-angular/plugins/karma')
    ],
    client: {
      jasmine: {
        // you can add configuration options for Jasmine here
        // the possible options are listed at https://jasmine.github.io/api/edge/Configuration.html
        // for example, you can disable the random execution with `random: false`
        // or set a specific seed with `seed: 4321`
      },
      clearContext: false // leave Jasmine Spec Runner output visible in browser
    },
    jasmineHtmlReporter: {
      suppressAll: true // removes the duplicated traces
    },
    coverageReporter: {
      dir: require('path').join(__dirname, './coverage/book-monkey'),
      subdir: '.',
      reporters: [
        { type: 'html' },
        { type: 'text-summary' }
      ]
    },
    reporters: ['progress', 'kjhtml'],
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false,
    restartOnFileChange: true
  });
};

```

```
├── @stryker-mutator/core@6.1.2
├── @stryker-mutator/karma-runner@6.1.2
├── @stryker-mutator/typescript-checker@6.1.2

```

```
├── @stryker-mutator/karma-runner@6.1.2
├── karma-chrome-launcher@3.1.1
├── karma-coverage@2.2.0
├── karma-jasmine-html-reporter@1.7.0
├── karma-jasmine@5.0.1
├── karma@6.3.20

```

**Test runner environment**

```shell
 ng test
```

**Your Environment**

| software         | version(s)
| ---------------- | -------
| node             | v16.15.1
| npm              | 8.11.0
| Operating System | maxos 12.4

**Add stryker.log**

**Comment by bodote** [comment]
is there any way to suppress deleting the `.stryker-tmp ` after the `stryker run` so that we can see , whats going on there?

**Comment by bodote** [comment]
found a workaround : `--inPlace` prevents that bug from happening.

**Comment by bodote** [comment]
...and here is the log file
[stryker1.log](https://github.com/stryker-mutator/stryker-js/files/9411378/stryker1.log)

**Comment by bodote** [comment]
specifically the `--ignorePattern= ` does not work with the workaround mentioned above `--inPlace`. 
because the thing is, that the command line:
` stryker run  --ignorePatterns='**','!src/**/search.effects.ts' --fileLogLevel=trace` leads to strange type-errors , such as:
```
10:39:08 (87003) TRACE ChildProcessProxy 
Error: src/app/search/search.effects.spec.ts:46:26 - error TS2339: Property 'pipe' does not exist on type 'EffectResult<Action> & CreateEffectMetadata'.
  Property 'pipe' does not exist on type '((...args: any[]) => Observable<Action>) & CreateEffectMetadata'.
46     effects.searchBooks$.pipe(toArray()).subscribe((actions) => {
```
while changing  the stryker.conf.json option 
```
 "mutate": [
    "src/**/*.ts",
    "!src/**/*.spec.ts",
```
to : 
```
 "mutate": [
    "src/**/search.effects.ts",
    "!src/**/*.spec.ts",
```
runs on the same source file without any problems.
to reproduce use: https://github.com/bodote/book-monkey/commit/1af6a02e66d11aae45f0a42656970aad2fb5fb89

**Comment by bodote** [comment]
problem persists after update all npm packages

**Comment by bodote** [comment]
what I found out so far with the debugger: 
apearantly the content of the `--ignore-pattern` commandline option is already used here: 
```
resolveInputFileNames (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/fs/project-reader.ts:194)
await (Unknown Source:0)
read (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/fs/project-reader.ts:53)
execute (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/process/1-prepare-executor.ts:65)
processTicksAndRejections (internal/process/task_queues:95)
await (Unknown Source:0)
runMutationTest (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker.ts:30)
runMutationTest (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker-cli.ts:44)
run (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker-cli.ts:201)
run (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker-cli.ts:205)
<anonymous> (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/bin/stryker.js:6)
run (internal/modules/esm/module_job:193)
``` 
and added as rule 6 and 7 to the default `ignoreRules` which in turn leads to only one file left in the variable `project->files->Map` in
execute (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/process/1-prepare-executor.ts:67)

only after that the sandbox is filled with the only one file, selected by the `--ignorePatter`- option, which in turn leads to the problem that when the sandbox directory is filled, it is filled only with this only one file, but the configuration files from the root dir of the project to be analysed are therefore missing now in the sandbox-dir

When the dryRun is startet here:
```
runMutationTest (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker.ts:35)
await (Unknown Source:0)
runMutationTest (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker-cli.ts:44)
run (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker-cli.ts:201)
run (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/src/stryker-cli.ts:205)
<anonymous> (/Users/bodo.teichmann/dev/learning/myStryker/stryker-js/packages/core/bin/stryker.js:6)
run (internal/modules/esm/module_job:193)
```
stryker is complaining about the missing `stryker-karma.conf.js` file , since it is not in the sandbox directory. 

this also explains , why when stryker is run with the `--inPlace` option, this problem does not occur.

contrary then not using the `--ignorePattern` cmdline option at all, but instead selecting on file via the "mutate" parameter in `stryker.conf.json`, the sandbox is filled with all project files. the filtering to one file to be analyzed seems to be applied only after that . 

therefore the `--ignorePattern` seems to be NO replacement for the deprecated `--files` command line option.
the [documentation](https://stryker-mutator.io/docs/stryker-js/configuration/#ignorepatterns-string) however claims that 
`--files` is deprecated and   `--ignorePattern` should be used instead to run stryker with only a selected set of files

**Comment by bodote** [comment]
even if i try this: 
`stryker run --ignorePatterns='src/app/**','!src/app/search/search*.ts'` 
then i got this output:
``` 
[0834][bodo.teichmann@ba-mb-006:~/dev/learning/angular_buch/book-monkey]$ stryker run --ignorePatterns='src/app/**','!src/app/search/search*.ts'
08:35:21 (27116) WARN ProjectReader Glob pattern "!src/**/*.module.ts" did not exclude any files.
08:35:21 (27116) INFO ProjectReader Found 6 of 2364 file(s) to be mutated.
08:35:21 (27116) INFO Instrumenter Instrumented 6 source file(s) with 34 mutant(s)
08:35:22 (27116) INFO ConcurrencyTokenProvider Creating 2 checker process(es) and 2 test runner process(es).
08:36:31 (27116) INFO DryRunExecutor Starting initial test run (karma test runner with "perTest" coverage analysis). This may take a while.
08:36:31 (27116) INFO DryRunExecutor Initial test run succeeded. Ran 0 tests in 0 seconds (net 0 ms, overhead 179 ms).
08:36:31 (27116) ERROR Stryker No tests were executed. Stryker will exit prematurely. Please check your configuration.
```

**Comment by bodote** [comment]
instead `stryker run --mutate='src/**/search*.ts'` seems to work just fine.

- Which leads to the question, what is the "--ignorePatters" option good for anyway? 
- Why is "--ignorePatters" option recommended as a replacement for the deprecated `--files` option , when only `--mutate` seems to do the job ?

**Comment by bodote** [comment]
however `--mutate` used on the command line seems to **replace** all patters defined in `mutate` in the `stryker.conf.json` instead to complement those patterns. 
that leads to the following problem:
when I use `--mutate='src/**/search*.ts'` , stryker tries to also mutate and run the `*.spec.ts` files, which makes no sense.
Therefore you need to write : ` stryker run --mutate='src/**/search*.ts','!src/**/*.spec.ts'` in Order to reach the goal of running stryker agains only a subset of your project.

**Comment by bodote** [comment]
Summary of my findings :
TL;DR: the problem might be the documentation.

Because:

What I was trying to do is to mutate just one specific file in my project.
The documentation in `docs/configuration.md` (or https://stryker-mutator.io/docs/stryker-js/configuration/#file ) for the `--file` option says:
_DEPRECATED. Please use ignorePatterns instead, or use mutate to select which files to mutate._

However `stryker run --ignorePatterns='**','!src/**/search.effects.ts'` did not work as described in [this Bug 3688](https://github.com/stryker-mutator/stryker-js/issues/3688)

After debugging a little, its clear that the usage of
`stryker run --ignorePatterns='**','!src/**/search.effects.ts' ` on this project `https://github.com/bodote/book-monkey/commit/2a8143dd8506494e57034c7f9a22b073fff2fd3d` will not work because the `--ignorePattern` is applied also to the project files **while** copying these files to the sandbox. Therefore also the `karma.config.js` is missing and the `src/**/search.effects.spec.ts` are missing in the sandbox and in turn stryker fails with an error because of missing. (However there is a workaround that with `--inPlace`)
So I tried  
`stryker run --ignorePatterns='src/app/*','!src/**/search.e*.ts'`
which also does **not** work:

```
08:12:04 (18283) INFO ProjectReader Found 37 of 2452 file(s) to be mutated.
08:12:04 (18283) INFO Instrumenter Instrumented 37 source file(s) with 558 mutant(s)
08:12:04 (18283) INFO ConcurrencyTokenProvider Creating 2 checker process(es) and 2 test runner process(es).
08:12:41 (18283) INFO DryRunExecutor Starting initial test run (karma test runner with "perTest" coverage analysis). This may take a while.
08:12:42 (18283) INFO DryRunExecutor Initial test run succeeded. Ran 0 tests in 0 seconds (net 0 ms, overhead 210 ms).
08:12:42 (18283) ERROR Stryker No tests were executed. Stryker will exit prematurely. Please check your configuration.
```

then I tried
`stryker run --mutate='src/**/search.e_.ts'`
which, weirdly enough tries to also mutate the test file !:

```
Ran 32.42 tests per mutant on average.
------------------------|---------|----------|-----------|------------|----------|---------|
File                    | % score | # killed | # timeout | # survived | # no cov | # error |
------------------------|---------|----------|-----------|------------|----------|---------|
All files               |   53.49 |       23 |         0 |         18 |        2 |      10 |
 search.effects.spec.ts |   53.49 |       23 |         0 |         18 |        2 |       4 |
 search.effects.ts      |     n/a |        0 |         0 |          0 |        0 |       6 |
------------------------|---------|----------|-----------|------------|----------|---------|
```

Ok then, lets try --files nevertheless:
first :
`stryker run -f src/**/search.e*.ts`
Error:

```
08:20:02 (24095) ERROR Stryker Invalid config file "src/app/app.component.spec.ts". Error during import. Inner error: TypeError: ERR_UNKNOWN_FILE_EXTENSION (undefined) TypeError: Unknown file extension ".ts" for /Users/bodo.teichmann/dev/learning/angular_buch/book-monkey/src/app/app.component.spec.ts
    at new NodeError (node:internal/errors:393:5)
```

next trial:
`stryker run --files='src/**/search.e*.ts'`
again errors:

```
08:22:19 (25767) WARN OptionsValidator DEPRECATED. Use of "files" is deprecated, please use "ignorePatterns" instead (or remove "files" altogether will probably work as well). For now, rewriting them as ["**","!src/**/search.e*.ts"]. See https://stryker-mutator.io/docs/stryker-js/configuration/#ignorepatterns-string
08:22:19 (25767) WARN ProjectReader Glob pattern "!src/test.ts" did not exclude any files.
08:22:19 (25767) WARN ProjectReader Glob pattern "!src/environments/*.ts" did not exclude any files.
08:22:19 (25767) WARN ProjectReader Glob pattern "!src/**/*.module.ts" did not exclude any files.
08:22:19 (25767) WARN ProjectReader Glob pattern "!src/main.ts" did not exclude any files.
08:22:19 (25767) WARN ProjectReader Glob pattern "!cypress/**" did not exclude any files.
08:22:19 (25767) INFO ProjectReader Found 1 of 2 file(s) to be mutated.
08:22:19 (25767) INFO Instrumenter Instrumented 1 source file(s) with 6 mutant(s)
08:22:19 (25767) INFO ConcurrencyTokenProvider Creating 2 checker process(es) and 2 test runner process(es).
08:22:25 (25767) INFO DryRunExecutor Starting initial test run (karma test runner with "perTest" coverage analysis). This may take a while.
08:22:25 (25767) INFO DryRunExecutor Initial test run succeeded. Ran 0 tests in 0 seconds (net 0 ms, overhead 54 ms).
08:22:25 (25767) ERROR Stryker No tests were executed. Stryker will exit prematurely. Please check your configuration.
```

So what do you do, if you want mutate just one specific file in your project?

Well I found a solution, but my point here is, that the documentation needs better examples like this one:
`stryker run --mutate='src/**/search.e*.ts','!src/**/*.spec.ts'`
which finally does what I wanted.

But before I make any suggestions to better documentation, I need to better understand.. :

- why is "--files" deprecated?
- what is the intention behind "--ignorePattern"
- what is the basic idea behind the "--mutate" and should it supplement or replace whats already in the stryker.conf.json of a project ?

**Comment by nicojs** [comment]
Wow I think this all could have been avoided with improved docs. 

> Why is `--files` deprecated.

You can do the same with' files' by using `ignorePatterns`. Stryker internally rewrites `files` to `ignorePatterns` when you run stryker with `--files`. The reason to deprecate it is that people were confused, thinking that you specify which files to mutate with `files`. 

> what is the intention behind "--ignorePattern".

The intention is to ignore big directories/files you might have in your repo that has nothing to do with your code. For example, 1.5GB of movie/image files. See the original PR for more details: https://github.com/stryker-mutator/stryker-js/pull/2848#issue-864152020

> what is the basic idea behind the "--mutate"

With `mutate` you specify which files to mutate.

> should those command line options supplement or replace whats already in the stryker.conf.json of a project ?

Command line arguments _always replace_ configuration options in "stryker.conf.json". 

Does this clarify it for you?

**Comment by bodote** [comment]
yes , thank you !
